console-subscriber/tests/support/subscriber.rs

```diff
@@ -0,0 +1,323 @@
+use std::{collections::HashMap, fmt, future::Future, thread};
+
+use console_api::{
+    field::Value,
+    instrument::{instrument_client::InstrumentClient, InstrumentRequest},
+};
+use console_subscriber::ServerParts;
+use futures::stream::StreamExt;
+use tokio::{io::DuplexStream, task};
+use tonic::transport::{Channel, Endpoint, Server, Uri};
+use tower::service_fn;
+
+use super::state::{TestState, TestStep};
+use super::task::{ActualTask, ExpectedTask, TaskValidationFailure};
+
+pub(crate) const MAIN_TASK_NAME: &str = "console-test::main";
+const END_SIGNAL_TASK_NAME: &str = "console-test::signal";
+
+#[derive(Debug)]
+struct TestFailure {
+    failures: Vec<TaskValidationFailure>,
+}
+
+impl fmt::Display for TestFailure {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Task validation failed:\n")?;
+        for failure in &self.failures {
+            write!(f, " - {failure}\n")?;
+        }
+        Ok(())
+    }
+}
+
+/// Runs the test
+///
+/// This function runs the whole test. It sets up a `console-subscriber` layer
+/// together with the gRPC server and connects a client to it. The subscriber
+/// is then used to record traces as the provided future is driven to
+/// completion on a current thread tokio runtime.
+///
+/// This function will panic if the expectations on any of the expected tasks
+/// are not met or if matching tasks are not recorded for all expected tasks.
+#[track_caller]
+pub(super) fn run_test<Fut>(expected_tasks: Vec<ExpectedTask>, future: Fut)
+where
+    Fut: Future + Send + 'static,
+    Fut::Output: Send + 'static,
+{
+    use tracing_subscriber::prelude::*;
+
+    let (client_stream, server_stream) = tokio::io::duplex(1024);
+    let (console_layer, server) = console_subscriber::ConsoleLayer::builder().build();
+    let registry = tracing_subscriber::registry().with(console_layer);
+
+    let mut test_state = TestState::new();
+    let mut test_state_test = test_state.clone();
+
+    let join_handle = thread::Builder::new()
+        .name("console::subscriber".into())
+        .spawn(move || {
```

Comment:
i was going to leave a comment asking why we needed to spawn this stuff in a background thread, but i realized that it's to ensure that any `tracing` spans emitted by the client and console server tasks aren't recorded by the test console subscriber. can we maybe include a comment explaining this? something like:
```suggestion
        // Run the test's console server and client tasks in a separate thread
        // from the main test, ensuring that any `tracing` emitted by the
        // console worker and the client are not collected by the subscriber
        // under test.
        .spawn(move || {
```